deps/chakrashim/lib/chakra_debug.js

```diff
@@ -0,0 +1,1304 @@
+                               responseJson);
+              chakraDebug.SendDelayedRespose(responseJson);
+            } else {
+              unResolvedBreakpoints.push(pendingBreakpointRequest);
+            }
+          } catch (ex) {
+            Logger.LogError('ProcessPendingBreakpoints exception: ' +
+                           ex.message,
+                           ex.stack);
+          }
+        }
+        Array.prototype.push.apply(_pendingBreakpoints, unResolvedBreakpoints);
+      },
+      RemoveBreakpoint: function(id) {
+        var v8Breakpoint = _breakpoints[id];
+        if (v8Breakpoint) {
+          v8Breakpoint.Remove();
+          delete _breakpoints[id];
+        }
+      }
+    };
+
+    var _scripts = [];
+    var _scriptHandles = [];
+
+    var ScriptsManager = {
+      GetScripts: function() {
+        var scriptsArray = CallHostFunctionNoLog(chakraDebug.JsDiagGetScripts);
+        scriptsArray.forEach(function(element, index, array) {
+          if (!_scripts[element.scriptId]) {
+            var scriptHandle = GetNextInternalHandle();
+            _scriptHandles[scriptHandle] = scriptHandle;
+            _scripts[element.scriptId] = new V8Script(element, scriptHandle);
+          }
+        });
+        return _scripts;
+      },
+      GetScript: function(scriptId) {
+        if (!_scripts[scriptId]) {
+          this.GetScripts();
+        }
+        return _scripts[scriptId];
+      },
+      IsScriptHandle: function(handle) {
+        return (_scriptHandles[handle]) ? true : false;
+      },
+      GetScriptFromHandle: function(handle) {
+        for (var i = 0; i < _scripts.length; ++i) {
+          if (_scripts[i] && _scripts[i].GetHandle() == handle) {
+            return _scripts[i];
+          }
+        }
+      },
+      GetScriptFromFileName: function(fileName) {
+        var scripts = this.GetScripts();
+        for (var i = 0; i < scripts.length; ++i) {
+          if (scripts[i] && scripts[i].IsSameFileName(fileName)) {
+            return scripts[i];
+          }
+        }
+      },
+      GetScriptFromMatchingFileName: function(regexToMatch) {
+        var scripts = this.GetScripts();
+        for (var i = 0; i < scripts.length; ++i) {
+          if (scripts[i] && scripts[i].GetFileName() &&
+            scripts[i].GetFileName().match(regexToMatch)) {
+            return scripts[i];
+          }
+        }
+      }
+    };
+
+    return {
+      Utility: Utility, BreakpointManager: BreakpointManager,
+      ScriptsManager: ScriptsManager
+    };
+  })();
+
+
+  var ExecutionStateType = {
+    Break: 0,
+    Async: 1
+  };
+
+  function ExecutionState(stateType, scriptId) {
+    this.type = stateType;
+    this.breakScriptId = scriptId;
+    this.frames = undefined;
+    this.objectEval = undefined;
+    this.objectProtoEval = undefined;
+    this.objectPrototypeEval = undefined;
+  }
+
+  ExecutionState.prototype.ChangeToBreak = function() {
```

Comment:
Most of these is trying to mimic how its done in V8 (debug.js) and I would like to keep it as close as possible in terms of structure.

---
In reply to: [82246479](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82246479) [](ancestors = 82246479)